app/db.py

```diff
@@ -0,0 +1,501 @@
+    print("Found DOI's: {0}".format(len(list_dois)))
+
+    for i in range(0, len(list_dois), batch_size):
+        db_urls_added = []
+        db_requests_added = []
+        for row in tqdm(list_dois[i : i + batch_size]):
+            url_resp_dict = {}
+            url = "https://api.unpaywall.org/v2/{0}?email={1}".format(
+                quote(row.doi), email
+            )
+            resp = request_unpaywall_api(url)
+            resp_data = resp.json()
+            req_dict = {
+                "doi": row.doi,
+                "request_url": url,
+                "request_type": "unpaywall",
+                "response_content": dumps(resp_data),
+                "response_status": resp.status_code,
+            }
+            db_requests_added.append(req_dict)
+            row.url_unpaywall = True
+
+            # check if response includes needed data
+            if "doi_url" in resp_data:
+                url_resp_dict["unpaywall:doi"] = resp_data["doi_url"]
+            if "oa_locations" in resp_data:
+                for loc in resp_data["oa_locations"]:
+                    if "url_for_pdf" in loc:
+                        if loc["url_for_pdf"]:
+                            url_resp_dict["unpaywall:pdf"] = loc["url_for_pdf"]
+                    if "url" in loc:
+                        if loc["url"]:
+                            url_resp_dict["unpaywall:url"] = loc["url"]
+                    if "url_for_landing_page" in loc:
+                        if loc["url_for_landing_page"]:
+                            url_resp_dict["unpaywall:landing_page"] = loc[
+                                "url_for_landing_page"
+                            ]
+
+            # store URL's in database
+            for url_type, url in url_resp_dict.items():
+                if url not in url_list + urls_added:
+                    url_dict = {"url": url, "doi": row.doi, "url_type": url_type}
+                    urls_added.append(url)
+                    db_urls_added.append(url_dict)
+        create_entities(db, Request, db_requests_added)
+        create_entities(db, Url, db_urls_added)
+        num_urls_unpaywall_added += len(db_urls_added)
+        num_requests_added += len(db_requests_added)
+    print('{0} "Unpaywall" URL\'s added to database.'.format(num_urls_unpaywall_added))
+    print("{0} Requests added to database.".format(num_requests_added))
+
+
+def get_fb_data() -> None:
+    """Get app access token.
+
+    Example Response:
+    {'id': 'http://dx.doi.org/10.22230/src.2010v1n2a24', 'engagement': { 'share_count': 0, 'comment_plugin_count': 0, 'reaction_count': 0, 'comment_count': 0}}
+    """
+    config = get_config()
+    num_fbrequests_added = 0
+    app_id = config.FB_APP_ID
+    app_secret = config.FB_APP_SECRET
+    batch_size = config.FB_BATCH_SIZE
+
+    db = get_db()
+    token = get_graph_api_token(app_id, app_secret)
+    fb_graph = get_graph_api(token["access_token"], version="3.1")
+
+    result_url = get_all(db, Url)
```

Comment:
This is currently retrieving the same 100 URLs (skip=0, limit=100) for each call. There is obviously some logic missing here because of the scheduler, but I just wanted to point out that even without the scheduling/prioritization logic the question of getting a set of candidate URLs would probably need to be resolved.